### ISSUE TITLE
Potential false positive for invalid_runtime_symbol_definitions lint

### ISSUE THREAD

**Issue by hkalbasi** [rust-lang/rust#159446]
I tried this code:

```rust
unsafe extern "C" {
    fn exit(s: i32);
}

fn main() {
    unsafe { exit(5); }
}
```

I expected to see this happen: Compilation succeed and program returned 5

Instead, this happened:
```
error: invalid definition of the runtime `exit` symbol used by the standard library
 --> <source>:4:5
  |
4 |     fn exit(s: i32);
  |     ^^^^^^^^^^^^^^^^
  |
  = note: expected `unsafe extern "C" fn(i32) -> !`
          found    `unsafe extern "C" fn(i32)`
  = help: either fix the signature or remove any attributes like `#[unsafe(no_mangle)]`, `#[unsafe(export_name = "exit")]`, or `#[link_name = "exit"]`
  = note: `#[deny(invalid_runtime_symbol_definitions)]` on by default

error: aborting due to 1 previous error; 2 warnings emitted

Compiler returned: 1
```
It might be suspicious, but it is not invalid definitely.

### Meta
I hit this bug in [CO2](https://github.com/hkalbasi/co2) which is basically a C frontend for rustc. I would like to enable this very useful lint at least warn by default, but I don't want to annoy C people by forcing them to add `__atribute__((no_return))` to their `void exit(int);` declarations. I expect c2rust to have a similar problem.

cc @Urgau #155521

**Comment by Urgau** [comment]
POSIX.1-2024 and the "ISO C standard" [documents](https://pubs.opengroup.org/onlinepubs/9799919799/functions/exit.html) that "The `exit()` function does not return."

So the correct representation is condition is to use the never type as the return type.

As for whenever it's correct for it to be UB, I think it is, as least for C/C++, [n2764](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2764.pdf) *(couldn't find a more recent draft)* clear states that:
> If a function is declared with the noreturn attribute in one translation unit and the same function is declared without the noreturn attribute in another translation unit, the behavior is undefined.

musl [defines `exit` with the `_Noreturn` attribute](https://github.com/kraj/musl/blob/a42e9dee266f398026a33d0793c66225c7997755/src/exit/exit.c#L30) (same attribute but for C).

I don't know if we have such wording in Rust (at least it's not immediately obvious in [Behavior considered undefined](https://doc.rust-lang.org/stable/reference/behavior-considered-undefined.html#behavior-considered-undefined)).

cc @RalfJung in case I missed something.

**Comment by hkalbasi** [comment]
Clearly exit should exit the program, but there is no problem in a declaration of `exit` which doesn't return `!` and return `()`. It would be a weaker declaration with smaller guarantee. Similarly, I think this is not UB:
```rust
#[unsafe(no_mangle)]
fn exit() { // Note the lack of -> !
    loop {} // Or another non termination
}
```
Since `!` and `()` are ABI compatible. So I think this lint for `!` and `()` belongs to the `suspicious_runtime_symbol_definitions`, not `invalid_runtime_symbol_definitions`.

**Comment by Urgau** [comment]
> there is no problem in a declaration of `exit` which doesn't return `!` and return `()`.

At least in C/C++ world it doesn't look like it is (see [n2764](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2764.pdf) and my previous message).

**Comment by hkalbasi** [comment]
I don't see how "The exit() function does not return." conflicts with "`_noReturn` is not required on the exit declaration". Are you claiming that this C23 code has UB?
```C
void exit(int status);

int main() {
    exit(5);
}
```
And the standard requires the declaration to be `void _Noreturn exit(int status);`? I would be surprised if this is the case.

If that C code is UB free, then I expect the Rust equivalent code in the OP to be UB free too. So that code, while not ideal, is not "invalid".

**Comment by RalfJung** [comment]
> Since ! and () are ABI compatible. 

So there is no ABI where `noreturn` does any kind of shenanigans?
And even CFI treats them as compatible?
Cc @rust-lang/opsem @Darksonn

**Comment by Urgau** [comment]
I don't have access to the C standard so this may not be accurate, but from reading [n2764](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2764.pdf) "Semantics 3." section, to me the answer is yes to both of your questions.

It is by the reading of [n2764](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2764.pdf) UB to define `exit` without `_Noreturn`.

**Comment by RalfJung** [comment]
The C standard (more precisely, a draft that's fully identical to the standard except for legal purposes) is available in freely (and legally) online, [here's the latest version](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3096.pdf).

**Comment by RalfJung** [comment]
Yeah C has lots of silly UB around declarations. But we do document `()` and `!` as ABI-compatible and even @Jules-Bertholet's PR doesn't change that, so on the Rust side we *should* be able to accept this.

**Comment by Urgau** [comment]
Okay, so the declaration is fine and shouldn't be linted. What about those implementation?
```rust
#[no_mangle]
extern "C" fn exit(s: i32) {} // is this UB?

#[no_mangle]
extern "C" fn exit(s: i32) { loop{} } // what about this one?
```

**Comment by RalfJung** [comment]
That `no_mangle` is unsafe. So the first one is UB for the same reason that this is UB
```rust
#[no_mangle]
extern "C" fn write(...) -> ... { 0 }
```
When you define a `no_mangle` function with a well-known name you have to implement the contract associated with that name. This has nothing to do with the signature.

The 2nd one is not UB.

**Comment by Darksonn** [comment]
I put together a godbolt snippet to check the CFI types with `-> !`, which you may find [here][1]. It appears that they are different CFI types right now:

* `_ZTSFvvE`
* `_ZTSFu5nevervE`

However this mismatch is entirely up to Rust and not something forced upon us by CFI. In the C language the CFI type of a noreturn function is the same as the CFI type of a function that returns `void`. We could simply change any Rust type with trivial ABI to use the same CFI name as returning void.

[1]: https://rust.godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,selection:(endColumn:2,endLineNumber:7,positionColumn:2,positionLineNumber:7,selectionStartColumn:2,selectionStartLineNumber:7,startColumn:2,startLineNumber:7),source:'pub+unsafe+extern+%22C%22+fn+foo()+%7B%0A++++%0A%7D%0A%0Apub+unsafe+extern+%22C%22+fn+foo2()+-%3E+!!+%7B%0A++++loop+%7B%7D%0A%7D'),l:'5',n:'0',o:'Rust+source+%231',t:'0')),k:47.75541795665635,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:nightly,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'1',execute:'1',intel:'0',libraryCode:'0',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,libs:!(),options:'-Zsanitizer%3Dcfi+-Clto+-Cunsafe-allow-abi-mismatch%3Dsanitizer+--emit%3Dllvm-ir+-Cdebuginfo%3D0',overrides:!((name:edition,value:'2024')),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+rustc+nightly+(Editor+%231)',t:'0'),(h:output,i:(compilerName:'rustc+1.95.0',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+rustc+nightly+(Compiler+%231)',t:'0')),k:52.244582043343655,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4

**Comment by Urgau** [comment]
Opened #159513 to consider it as suspicious only (so warn-by-default).

**Comment by hkalbasi** [comment]
@Darksonn Does it mean that Rust code can't pass `extern "C" fn() -> !` function pointer to C code in a CFI enabled setup, in the current state of things? (Is CFI relevant at all in cross language setups? I don't know much about CFI) If so, I guess it is not very ideal.
